server/src/main/java/io/unitycatalog/server/service/PermissionService.java

```diff
@@ -0,0 +1,278 @@
+      "#authorize(#principal, #metastore, OWNER) || #authorize(#principal, #registered_model, OWNER)")
+  @AuthorizeKey(METASTORE)
+  public HttpResponse updateRegisteredModelAuthorization(
+      @Param("name") @AuthorizeKey(REGISTERED_MODEL) String name, UpdatePermissions request) {
+    return updateAuthorization(REGISTERED_MODEL, name, request);
+  }
+
+  private HttpResponse updateAuthorization(
+      SecurableType securableType, String name, UpdatePermissions request) {
+    UUID resourceId = getResourceId(securableType, name);
+    List<PermissionsChange> changes = request.getChanges();
+    Set<UUID> principalIds = new HashSet<>();
+    changes.forEach(
+        change -> {
+          String principal = change.getPrincipal();
+          User user = USER_REPOSITORY.getUserByEmail(principal);
+          UUID principalId = UUID.fromString(Objects.requireNonNull(user.getId()));
+          principalIds.add(principalId);
+          change
+              .getAdd()
+              .forEach(
+                  privilege ->
+                      //  Privileges should always be a superset of Privilege so this _should_
+                      // always be non-null but let's be safe anyway.
+                      Optional.ofNullable(Privileges.fromPrivilege(privilege))
+                          .map(p -> authorizer.grantAuthorization(principalId, resourceId, p)));
+          change
+              .getRemove()
+              .forEach(
+                  privilege ->
+                      //  Privileges should always be a superset of Privilege so this _should_
+                      // always be non-null but let's be safe anyway.
+                      Optional.ofNullable(Privileges.fromPrivilege(privilege))
+                          .map(p -> authorizer.revokeAuthorization(principalId, resourceId, p)));
+        });
+
+    Map<UUID, List<Privileges>> authorizations = authorizer.listAuthorizations(resourceId);
+    List<PrivilegeAssignment> privilegeAssignments =
+        authorizations.entrySet().stream()
+            .filter(entry -> principalIds.contains(entry.getKey()))
+            .map(
+                entry -> {
+                  List<Privilege> privileges =
+                      entry.getValue().stream()
+                          .map(Privileges::toPrivilege)
+                          // mapping to Privilege may result in nulls since Privilege is a subset of
+                          // Privileges, so filter them out.
+                          .filter(Objects::nonNull)
+                          .toList();
+                  return new PrivilegeAssignment()
+                      .principal(USER_REPOSITORY.getUser(entry.getKey().toString()).getEmail())
+                      .privileges(privileges);
+                })
+            .filter(assignment -> !assignment.getPrivileges().isEmpty())
+            .collect(Collectors.toList());
+
+    return HttpResponse.ofJson(new PermissionsList().privilegeAssignments(privilegeAssignments));
```

Comment:
This code overlaps significantly with lines 121-145 -- can it be refactored into a separate method that takes the principal ids to filter for?